File: 138/copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
//需要 HashMap 幫忙對應 old to new
//然後走2輪, 第1輪建map, 第2輸照著map建random連結
class Solution {
    HashMap<Node,Node> map = new HashMap<>();
    public Node copyRandomList(Node head) {
        if(head==null) return null;
        //case: 3/19: [] 空指標
        Node ans = new Node(head.val);
        map.put(head, ans);

        Node next = head.next;
        Node next2 = ans;
        while(next != null) {
            Node temp = new Node(next.val);
            map.put(next, temp);
            next2.next = temp;
            next2 = next2.next;
            next = next.next;
        }//第一輪,照著值建出結構

        next = head;
        next2 = ans; //第二輪, 查 map 確認 random值
        while(next != null){
            next2.random = map.get(next.random);
            next2 = next2.next;
            next = next.next;
        }
        return ans;
    }
}

File: 138/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
# 使用 HashMap 來找到 Random pointer 的對應關係
# two-pass，第一次建出連接架構 & 建 hash table
# 第二次（查表）填好random資訊
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        table = {} # table[舊] = 新
        ans = Node(0) # 從 ans.next 開始放值
        head2 = head # 不想破壞 head, 所以複製成 head2
        now = ans
        while head2 != None:
            now.next = Node(head2.val)
            table[head2] = now.next # 最重要，建立「舊-新關係」
            head2 = head2.next
            now = now.next

        head2 = head
        now = ans.next
        while head2 != None:
            if head2.random == None:
                now.random = None # 因前面漏建 table[None] = None
            else:
                now.random = table[head2.random]
            now = now.next
            head2 = head2.next
        return ans.next
